1.Add the machine-provider-nutanix types 2.Add fields to Infrastructure NutanixPlatformSpec

yanhua121 opened the pull request:
1.Add the machine-provider-nutanix types 
2.Add the Nutanix PC endpoint/port fields to Infrastructure NutanixPlatformSpec

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -705,7 +705,15 @@ type AlibabaCloudResourceTag struct {
-type NutanixPlatformSpec struct{}
+type NutanixPlatformSpec struct {
+	// pcEndpoint is the endpoint (DNS name or IP address) of the Nutanix Prism Central
+	// +required
```
I think the correct format for this is slightly different, looking at the generated CRD the field hasn't been marked as required
```suggestion
	// +kubebuilder:validation:Required
```

yanhua121 replied:
I see lots of uses of "// +required" in this repo (for example, in the "config/" folder). What's the difference of these two format?  The marker "// +required" does not generate "required" marker for the field in CRD?

JoelSpeed replied:
The format I've suggested will add the field to the required fields within the OpenAPI schema, the current format isn't actually recognised by the generator, so doesn't add anything to the OpenAPI schema.

So with the current tag, it doesn't do anything, it will actually make a change with the new tax 

This is semi trivial because the MachineProviderConfig isn't actually a real CRD, but we have plans to generate documentation from OpenAPI schemas so it would be good to have the correct tags

yanhua121 replied:
I see, will fix these.

yanhua121 replied:
Fixed.

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+	"k8s.io/klog/v2"
+)
+
+const (
+	// NutanixCredentialsSecretName is the name of the secret holding the credentials for PC client
+	NutanixCredentialsSecretName = "nutanix-credentials"
+)
```
Does this need to be here? I wouldn't expect this to be here based on what we have in other platforms. What's the plan for use of this const?

yanhua121 replied:
Will remove this

yanhua121 replied:
Removed.

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// clusterReferenceUuid is the UUID of the PE/cluster the Machine's VM will be created
```
I think this might be missing a word or two, `the Machine's VM will be created in` maybe?

Is this field optional or required, is there any format for this?

yanhua121 replied:
yes, will add the word "in".

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+	// clusterReferenceUuid is the UUID of the PE/cluster the Machine's VM will be created
+	ClusterReferenceUUID string `json:"clusterReferenceUuid,omitempty"`
+
+	// imageUuid is the UUID of the rhcos image uploaded to the PC.
+	// If the imageUUID is configured in the Machine CR, it will be used to create the VM.
+	// Otherwise, the imageName will be used to obtain the imageUUID, before creating the VM.
+	ImageUUID string `json:"imageUuid,omitempty"`
+
+	// imageName is the name of the rhcos image uploaded to the PC
+	ImageName string `json:"imageName,omitempty"`
```
These should probably be in a struct called `NutanixImageReference` with a UUID or Name fields as a discriminated union type

yanhua121 replied:
OK, good suggestion. Will change.

yanhua121 replied:
Changed.

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+	// numSockets is the number of sockets of the VM to create
+	NumSockets int64 `json:"numSockets,omitempty"`
+
+	// memorySizeMib is the memory size in megabytes of the VM to create
+	MemorySizeMib int64 `json:"memorySizeMib,omitempty"`
+
+	// diskSizeMib is the disk size in megabytes of the VM to create
+	DiskSizeMib int64 `json:"diskSizeMib,omitempty"`
```
Have you considered using a `resource.Quantity` for these types so that users can specify a string capacity, eg, `120Gi` instead of `120*1024` whatever that happens to be?

yanhua121 replied:
Since Nutanix PC API just uses "Mib" size format, we want to keep consistent.

JoelSpeed replied:
For OpenShift, we prefer to make things more user friendly than adding large, non-human readable integers. The conversion within the controller should be relatively straightforward once the value is parsed, I would much rather see this as a quantity value to make it human readable

yanhua121 replied:
Changed as suggestion.

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+	// diskSizeMib is the disk size in megabytes of the VM to create
+	DiskSizeMib int64 `json:"diskSizeMib,omitempty"`
+
+	// powerState is the expected power state of the VM to create
+	PowerState string `json:"powerState,omitempty"`
```
Do we want users to be able to create VMs that are powered off? Does that create issues for OpenShift, eg, a Machine is created, but is never powered on. The Machine sits now in a `Provisioned` phase with no Node for a long time, this would trigger an alert to customers based on our current configuration of alerts

yanhua121 replied:
Initially we have this field for cluster-api MachinePool use case. It seems we don't need other VM "powerState" than "ON" here with Openshift. I will remove this field.

yanhua121 replied:
Removed.

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+	// conditions is a set of conditions associated with the Machine to indicate
+	// errors or other status
+	// +optional
+	Conditions []NutanixMachineProviderCondition `json:"conditions,omitempty"`
```
Do we need a new type here or can we just use a [metav1.Condition](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#Condition)?

Also, stylistic convention is to have the `Conditions` as the first object in the struct

yanhua121 replied:
Changed to use metav1.Condition

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+
+	// ready is true when the provider resource is ready.
+	// +optional
+	Ready bool `json:"ready"`
```
Bools are not allowed in Kubernetes style APIs, this would be better represented as a condition

yanhua121 replied:
Ok, I will integrate the VM "ready" state info conditions.

yanhua121 replied:
Removed this status field

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+// +openshift:compatibility-gen:level=1
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type NutanixMachineProviderStatus struct {
+	metav1.TypeMeta `json:",inline"`
```
Do we need typemeta here?

yanhua121 replied:
Not really, at least we don't use it in mapi-nutanix. I will remove it. But I see many other providers (alibaba, aws, etc.) use TypeMeta/ObjectMeta in their MachineProviderStatus struct.

JoelSpeed replied:
Ahh yes, actually, I think this might be needed. We embed the providerstatus within a rawextension, which I believe has a tag on it that checks for type meta within the resources, will double check that

yanhua121 replied:
okay

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+	CredentialsSecret *corev1.LocalObjectReference `json:"credentialsSecret,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NutanixMachineProviderConfigList contains a list of NutanixMachineProviderConfig
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type NutanixMachineProviderConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []AlibabaCloudMachineProviderConfig `json:"items"`
+}
```
This isn't going to be transformed into a CRD so I don't think we need a list type

yanhua121 replied:
Ok, will remove this. I see Alibaba has the ConfigList type.

JoelSpeed replied:
Alibaba shouldn't have one either, that was an omission on our part

yanhua121 replied:
Removed

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+	Conditions []NutanixMachineProviderCondition `json:"conditions,omitempty"`
+}
+
+// NutanixMachineProviderConditionType is a valid value for NutanixMachineProviderCondition.Type
+type NutanixMachineProviderConditionType string
+
+// Valid conditions for an Nutanix VM instance.
+const (
+	// NutanixMachineCreation indicates whether the machine's VM has been created or not. If not,
+	// it should include a reason and message for the failure.
+	NutanixMachineCreation NutanixMachineProviderConditionType = "MachineCreation"
+
+	// NutanixMachineUpdate indicates whether the machine's VM has been updated or not. If not,
+	// it should include a reason and message for the failure.
+	NutanixMachineUpdate NutanixMachineProviderConditionType = "MachineUpdate"
+
+	// NutanixMachineDeletion indicates whether the machine's VM has been deleted or not. If not,
+	// it should include a reason and message for the failure.
+	NutanixMachineDeletion NutanixMachineProviderConditionType = "MachineDeletion"
+)
+
+// NutanixMachineProviderConditionReason is reason for the condition's last transition.
+type NutanixMachineProviderConditionReason string
+
+const (
+	// NutanixMachineCreationSucceeded indicates machine creation success.
+	NutanixMachineCreationSucceeded NutanixMachineProviderConditionReason = "MachineCreationSucceeded"
+	// NutanixMachineCreationFailed indicates machine creation failure.
+	NutanixMachineCreationFailed NutanixMachineProviderConditionReason = "MachineCreationFailed"
+	// NutanixMachineCreationSucceeded indicates machine update success.
+	NutanixMachineUpdateSucceeded NutanixMachineProviderConditionReason = "MachineUpdateSucceeded"
+	// NutanixMachineCreationFailed indicates machine update failure.
+	NutanixMachineUpdateFailed NutanixMachineProviderConditionReason = "MachineUpdateFailed"
+	// NutanixMachineCreationSucceeded indicates machine deletion success.
+	NutanixMachineDeletionSucceeded NutanixMachineProviderConditionReason = "MachineDeletionSucceeded"
+	// NutanixMachineCreationFailed indicates machine deletion failure.
+	NutanixMachineDeletionFailed NutanixMachineProviderConditionReason = "MachineDeletionFailed"
+)
```
We already have a number of common condition types in the v1beta1 package, it may be better to keep generic condition types and reasons rather than creating specific Nutanix ones

yanhua121 replied:
I raised this question at the slack channel, see https://coreos.slack.com/archives/C0211848DBN/p1644522435725449. Since currently these common condition types in the v1beta1 package, it's not good to cross-reference those from the v1 package. We have to use the "Nutanix"-prefixed condition types and constants here, unless RH move those common condition types from  the "machine/v1beta1" package to a none version-specific package, or duplicate those in the "machine/v1" package.

yanhua121 replied:
Moved these constants to the mapi-nutanix repo. And use the upstream metav1.Condition type.

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+	var rawBytes []byte
+	var err error
+	if rawBytes, err = json.Marshal(status); err != nil {
+		return nil, fmt.Errorf("error marshalling providerStatus: %v", err)
+	}
+
+	return &runtime.RawExtension{
+		Raw: rawBytes,
+	}, nil
+}
+
+// NutanixMachineProviderSpecFromRawExtension unmarshals a raw extension into an NutanixMachineProviderConfig type
+func NutanixMachineProviderSpecFromRawExtension(rawExtension *runtime.RawExtension) (*NutanixMachineProviderConfig, error) {
+	if rawExtension == nil {
+		return &NutanixMachineProviderConfig{}, nil
+	}
+
+	spec := new(NutanixMachineProviderConfig)
+	if err := json.Unmarshal(rawExtension.Raw, &spec); err != nil {
+		return nil, fmt.Errorf("error unmarshalling providerSpec: %v", err)
+	}
+
+	klog.V(5).Infof("Got provider Spec from raw extension: %+v", spec)
+	return spec, nil
+}
+
+// NutanixMachineProviderStatusFromRawExtension unmarshals a raw extension into an NutanixMachineProviderStatus type
+func NutanixMachineProviderStatusFromRawExtension(rawExtension *runtime.RawExtension) (*NutanixMachineProviderStatus, error) {
+	if rawExtension == nil {
+		return &NutanixMachineProviderStatus{}, nil
+	}
+
+	providerStatus := new(NutanixMachineProviderStatus)
+	if err := json.Unmarshal(rawExtension.Raw, providerStatus); err != nil {
+		return nil, fmt.Errorf("error unmarshalling providerStatus: %v", err)
+	}
+
+	klog.V(5).Infof("Got provider Status from raw extension: %+v", providerStatus)
+	return providerStatus, nil
+}
```
Code like this is not allowed in the API repo, this must be moved into your machine-api-provider repository

yanhua121 replied:
Okay, will move these util functions to the mapi-nutanix repo.

yanhua121 replied:
Moved to mapi-nutanix repo.

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+	// imageName is the name of the rhcos image uploaded to the PC
+	ImageName string `json:"imageName,omitempty"`
+
+	// subnetUuid is the UUID of the network subnet to use for the Machine's VM
+	SubnetUUID string `json:"subnetUuid,omitempty"`
```
What's the possible format of this field? Is it required?

yanhua121 replied:
It's required field. I will add the marker comment to all required fields of the Spec.

yanhua121 replied:
Done

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+	// subnetUuid is the UUID of the network subnet to use for the Machine's VM
+	SubnetUUID string `json:"subnetUuid,omitempty"`
+
+	// numVcpusPerSocket is the number of vCPUs per socket of the VM to create
+	NumVcpusPerSocket int64 `json:"numVcpusPerSocket,omitempty"`
+
+	// numSockets is the number of sockets of the VM to create
+	NumSockets int64 `json:"numSockets,omitempty"`
+
+	// memorySizeMib is the memory size in megabytes of the VM to create
+	MemorySizeMib int64 `json:"memorySizeMib,omitempty"`
+
+	// diskSizeMib is the disk size in megabytes of the VM to create
+	DiskSizeMib int64 `json:"diskSizeMib,omitempty"`
```
Are these fields required? Do they have minimum or maximum values? Please add relevant kubebuilder validation tags

yanhua121 replied:
They are required fields. I will add the marker comment to all required fields of the Spec.
// +kubebuilder:validation:Required

yanhua121 replied:
Done

yanhua121 commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,244 @@
+
+	// vmUUID is the VM's UUID
+	// +optional
+	VmUUID *string `json:"vmUUID,omitempty"`
```
This field "vmUUID" is the UUID of the Nutanix VM created for the Machine CR. It uniquely identifies the VM instance. It has lots of use cases. One is when updating the Machine CR, the mapi-nutanix controller will use this vmUUID to make the nutanix PC API call to find the existing VM corresponds to the Machine CR, and then update the Machine CR's providerStatus fields, based on the VM's current status. Another use case is when deleting the Machine CR, the mapi-nutanix controller will use this vmUUID to make PC API call to delete the VM of the Machine.

yanhua121 replied:
Done

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+	// +kubebuilder:validation:Required
+	ImageReference ImageReference `json:"imageReference"`
+
+	// subnetUuid is the UUID of the cluster's network subnet to use for the Machine's VM
+	// +kubebuilder:validation:Required
+	SubnetUUID string `json:"subnetUuid"`
```
Is this the only way to identify a subnet or can they be looked up, for example via tags or name?

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+	// +kubebuilder:validation:Required
+	MemorySize resource.Quantity `json:"memorySize"`
+
+	// diskSize is the disk size (in Quantity format) of the VM to create
+	// +kubebuilder:validation:Required
+	DiskSize resource.Quantity `json:"diskSize"`
```
We will only support a single disk? Or is there a way to attach multiple disks?

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+
+// ImageReference holds the identity of the rhcos image uploaded to the PC
+type ImageReference struct {
+	// imageUuid is the UUID of the rhcos image uploaded to the PC.
+	// If the imageUUID is configured, it will be used to create the VM.
+	// Otherwise, the imageName will be used to obtain the imageUUID, before creating the VM.
+	ImageUUID string `json:"imageUuid,omitempty"`
+
+	// imageName is the name of the rhcos image uploaded to the PC
+	ImageName string `json:"imageName,omitempty"`
```
We can strip the `Image` prefixes off of these

yanhua121 replied:
changed

yanhua121 commented:
/retest

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// clusterReference is to identify the PE/cluster in which the Machine's VM will be created
+	// +kubebuilder:validation:Required
+	ClusterReference *NutanixReference `json:"clusterReference"`
```
What is `PE`? Can we spell that out please.
Are there any restrictions on this reference? Does it for example need to match any other values? Eg should it be matching the infrastructure object?

We don't need to include `Reference` in the name here, or any of the other references 

Are both UUID and Name suitable ways to look up a cluster?

I don't think this needs to be a pointer, can we make it a reference directly please
```suggestion
	// cluster is used to identify the P... E... /cluster in which the Machine's VM will be created.
	// The cluster UUID can be found ...
	// Alternatively you can identify the cluster by name.
	// +kubebuilder:validation:Required
	Cluster NutanixReference `json:"clusterReference"`
```

yanhua121 replied:
Addressed

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+	// +kubebuilder:validation:Required
+	ClusterReference *NutanixReference `json:"clusterReference"`
+
+	// imageReference is to identify the rhcos image uploaded to the Prism Central (PC)
+	// +kubebuilder:validation:Required
+	ImageReference *NutanixReference `json:"imageReference"`
```
This doesn't need to be a pointer (required fields shouldn't need to be pointers, so we can reduce nil pointer derefs later), and please drop the `Reference`

Are there any restrictions on this, is there any additional context we can add to the comment to help the user understand what information they need to gather for this? Can this be fetched by both Name and UUID?

Is there a particular service that the images are uploaded to or some nutanix specific terminology that might help a user understand who they might find the UUID/Name of the image?

yanhua121 replied:
Addressed

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+	// +kubebuilder:validation:Required
+	ImageReference *NutanixReference `json:"imageReference"`
+
+	// subnetReference is to identify the cluster's network subnet to use for the Machine's VM
+	// +kubebuilder:validation:Required
+	SubnetReference *NutanixReference `json:"subnetReference"`
```
Similar to above, we don't need a pointer or `Reference`.

If there's any additional details we can add to this to explain how the user might populate the values, that will help.
Are both Name and UUID valid ways to identify the subnet?

yanhua121 replied:
Addressed

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+	// +kubebuilder:validation:Required
+	SubnetReference *NutanixReference `json:"subnetReference"`
+
+	// numVcpusPerSocket is the number of vCPUs per socket of the VM to create
+	// +kubebuilder:validation:Required
+	NumVcpusPerSocket int64 `json:"numVcpusPerSocket"`
```
Are there any maximum or minimum values of this value?
Are there restrictions on this value imposed by using it in combination with other options at all?

I think this should probably be `numVCPUsPerSocket` but I'll get a second opinion on that capitalisation

JoelSpeed replied:
Wondering if the capacity configuration (CPUs, Sockets, Memory) should be in their own struct to separate them out from the rest of the flat structure

JoelSpeed replied:
I got a second opinion, let's go with `numVCPUsPerSocket` as you'd normally expect the acronym to be upper cased like this

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+	// +kubebuilder:validation:Required
+	NumVcpusPerSocket int64 `json:"numVcpusPerSocket"`
+
+	// numSockets is the number of sockets of the VM to create
+	// +kubebuilder:validation:Required
+	NumSockets int64 `json:"numSockets"`
```
Are there any maximum or minimum values of this?
Are there restrictions on this value imposed by using it in combination with other options at all?

I'm guessing this is to do with the CPU configuration, but the comment doesn't mention CPU at all, could we expand the comment a little to add a little more context about the effect this value has

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+	// +kubebuilder:validation:Required
+	NumSockets int64 `json:"numSockets"`
+
+	// memorySize is the memory size (in Quantity format) of the VM to create
+	// +kubebuilder:validation:Required
+	MemorySize resource.Quantity `json:"memorySize"`
```
Are there any maximum or minimum values for this?
Are there any restrictions imposed by using this in combination with other options?

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+	// +kubebuilder:validation:Required
+	MemorySize resource.Quantity `json:"memorySize"`
+
+	// diskSize is the system disk size (in Quantity format) of the VM to create
+	// +kubebuilder:validation:Required
+	DiskSize resource.Quantity `json:"diskSize"`
```
Are there any maximum or minimum values for this?
Are there any restrictions imposed by using this in combination with other options?

JoelSpeed replied:
Are there any other disk parameters that a user can set?

yanhua121 replied:
Since this field is in Quality format. It's hard to add kubebuilder:validation to enforce the size limit. So just added the minimum size in the doc.

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+	CredentialsSecret *corev1.LocalObjectReference `json:"credentialsSecret"`
+}
+
+// NutanixReference holds the identity of a Nutanix PC resource (cluster, image, subnet, etc.)
+type NutanixReference struct {
+	// uuid is the UUID of the resource in the PC.
+	// If this is configured, it will be used to create the VM.
+	// Otherwise, the resource name will be used to obtain the UUID, before creating the VM.
+	UUID string `json:"uuid,omitempty"`
+
+	// name is the resource name in the PC
+	Name string `json:"name,omitempty"`
+}
```
UUID and Name are mutually exclusive values right? This should be made into a [discriminated union](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1027-api-unions/README.md)

```suggestion
// NutanixReference holds the identity of a Nutanix PC resource (cluster, image, subnet, etc.)
// +union
type NutanixReference struct {
	// Type is the type of reference to use for this resource.
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Enum:="name";"uuid"
	Type NutanixReferenceType `json:"type"`
	
	// uuid is the UUID of the resource in the PC.
	// If this is configured, it will be used to create the VM.
	// Otherwise, the resource name will be used to obtain the UUID, before creating the VM.
	// +optional
	UUID *string `json:"uuid,omitempty"`

	// name is the resource name in the PC
	// +optional
	Name *string `json:"name,omitempty"`
}

// NutanixReferenceType is an enumeration of different resource reference types.
type NutanixReferenceType string

const (
	// NutanixUUIDReference is a resource reference referring to the object by UUID.
	NutanixUUIDReference NutanixReferenceType = "uuid"
	
	// NutanixNameReference is a resource reference referring to the object by Name.
	NutanixNameReference NutanixReferenceType = "name"
)
```

yanhua121 replied:
Applied your suggestion.

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+// NutanixReference holds the identity of a Nutanix PC resource (cluster, image, subnet, etc.)
+type NutanixReference struct {
+	// uuid is the UUID of the resource in the PC.
+	// If this is configured, it will be used to create the VM.
+	// Otherwise, the resource name will be used to obtain the UUID, before creating the VM.
```
We should make users select rather than inferring a preference. Kubernetes uses declarative APIs so inferring is not allowed by convention

yanhua121 replied:
Removed this comment.

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// vmState is the Machine associated VM's current state
+	// +optional
+	VmState *string `json:"vmState,omitempty"`
```
What will this value represent? Can we expand the comment here to identify what the user might expect to see within this field. What does it mean for it to be missing?

Does this need to be a pointer?

yanhua121 replied:
Removed this field. Will combine the vmState to the Conditions.

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,89 @@
+	// +optional
+	VmState *string `json:"vmState,omitempty"`
+
+	// vmUUID is the Machine associated VM's UUID
+	// +optional
+	VmUUID *string `json:"vmUUID,omitempty"`
```
Please expand the comment to explain what this field will be used for, what does it mean for it to be empty?

Does this need to be a pointer?

yanhua121 replied:
Added more doc to explain this field.

JoelSpeed reviewed (commented):
Hi, I've left a more thorough API review. In general I think this is looking good, but I've added a bunch of suggestions throughout. 

The main goal here is to make sure that the API is well documented, please review each of the comments on the new fields that you are adding and make sure that if an end user were to read them (that's the end goal here), that they would know exactly what the field is for, what values are expected, what happens if it's missing, are there any restrictions on the field etc

JoelSpeed reviewed (commented):
Do you happen to have a link to the Nutanix API docs that this project will interface with? It may help the review here to review the Nutanix API that we will be interfacing with

yanhua121 commented on config/v1/types_infrastructure.go:
```diff
@@ -705,7 +705,15 @@ type AlibabaCloudResourceTag struct {
+
+	// pcPort is the port of the Nutanix Prism Central
+	// +kubebuilder:validation:Required
+	PCPort string `json:"pcPort"`
```
Addressed.

yanhua121 commented on config/v1/types_infrastructure.go:
```diff
@@ -705,7 +705,15 @@ type AlibabaCloudResourceTag struct {
 // This only includes fields that can be modified in the cluster.
-type NutanixPlatformSpec struct{}
+type NutanixPlatformSpec struct {
+	// pcEndpoint is the endpoint (DNS name or IP address) of the Nutanix Prism Central
```
Addressed.

yanhua121 commented:
> interfacing

yanhua121 commented:
> 

Hi Joel, thanks for your thorough review and detailed comments and suggestions. I addressed all your comments. Please give another look at the code of this PR. We hope to merge this PR soon, as this PR is the basis of all the corresponding changes we have to make in the mapi-nutanix and installer repo.

For your reference, the Nutanix Prism Central API doc link is https://www.nutanix.dev/reference/prism_central/v3.

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,117 @@
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// vmUUID is the Machine associated VM's UUID
+	// The field is missing either before the VM is created, or after the VM is deleted
+	// Once the VM is created, the field is filled with the VM's UUID and it will not change
+	// until the VM is deleted. The vmUUID is used to find the VM when updating the Machine status,
+	// and to delete the VM when the Machine is deleted.
+	// +optional
+	VmUUID *string `json:"vmUUID,omitempty"`
```
Given the Machine maps 1:1 with a VM in Nutanix, I don't think we want to clear the UUID when the VM is deleted. What's the motivation behind removing the UUID from the status here? In other providers that do this, we add it, and then never remove/change it

yanhua121 replied:
There is a scenario that the Machine's associated VM is deleted not via deleting the Machine CR (but deleted from the PC Console or by calling the PC APIs). In this case, when the mapi-nutanix controller updating the Machine CR's status, it need to clear this vmUUID field, besides setting the corresponding status condition. In this scenario, the best solution would be for the Cloud Controller Manager (CCM) node controller to detect the node and VM is gone and then delete the corresponding Machine CR. Since Nutanix does not provide a CCM node controller yet, at least we need to make the Machine CR's status reflect the VM is deleted by clearing the vmUUID field and setting the corresponding status condition.

JoelSpeed replied:
In this scenario, I would expect the `Exists` call within the actuator to fail. If the machine is already provisioned or running, this moves the Machine into a failed phase. At which point, we also put a condition on the Machine to say that the instance no longer exists. 

I don't think we need this additional logic directly within the nutanix actuator, I would expect this field to stay populated.

Perhaps you can expand on what you're trying to achieve by removing the field value? Is something else relying on that?

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,117 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// clusterIdentifier is to identify the cluster (the Prism Element under management
+	// of the Prism Central), in which the Machine's VM will be created.
+	// The cluster identifier (uuid or name) can be obtained from the Prism Central console
+	// or using the prism_central API.
+	// +kubebuilder:validation:Required
+	ClusterIdentifier NutanixResourceIdentifier `json:"clusterIdentifier"`
+
+	// imageIdentifier is to identify the rhcos image uploaded to the Prism Central (PC)
+	// The image identifier (uuid or name) can be obtained from the Prism Central console
+	// or using the prism_central API.
+	// +kubebuilder:validation:Required
+	ImageIdentifier NutanixResourceIdentifier `json:"imageIdentifier"`
+
+	// subnetIdentifier is to identify the cluster's network subnet to use for the Machine's VM
+	// The cluster identifier (uuid or name) can be obtained from the Prism Central console
+	// or using the prism_central API.
+	// +kubebuilder:validation:Required
+	SubnetIdentifier NutanixResourceIdentifier `json:"subnetIdentifier"`
```
IMO we don't need any suffix on these fields, I would just make them `Cluster`, `Image` and `Subnet`, are you keen to have some sort of suffix on these? Typically we wouldn't have any suffix if you look at other provider APIs

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,117 @@
+	// +kubebuilder:validation:Required
+	SubnetIdentifier NutanixResourceIdentifier `json:"subnetIdentifier"`
+
+	// numVcpusPerSocket is the number of vCPUs per socket of the VM
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Minimum=1
+	NumVcpusPerSocket int32 `json:"numVcpusPerSocket"`
+
+	// numSockets is the number of sockets of the VM
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Minimum=1
+	NumSockets int32 `json:"numSockets"`
```
I think given the fields here are `int32`, we don't need the `num` prefix. I would suggest calling these `vcpusPerSocket` and `cpuSockets`, WDYT?

There was a previous thread on the name of the vcpus which I think you may have missed from the last review https://github.com/openshift/api/pull/1124#discussion_r822753752

JoelSpeed reviewed (commented):
This is looking a lot better, thanks for adding the additional context to the fields. I've added a few more naming suggestions, PTAL

JoelSpeed commented on machine/v1/types_nutanixprovider.go:
```diff
@@ -0,0 +1,117 @@
+	// +kubebuilder:validation:Required
+	Subnet NutanixResourceIdentifier `json:"subnet"`
+
+	// vcpusPerSocket is the number of vCPUs per socket of the VM
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Minimum=1
+	VcpusPerSocket int32 `json:"vcpusPerSocket"`
+
+	// vcpuSockets is the number of vCPU sockets of the VM
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Minimum=1
+	VcpuSockets int32 `json:"vcpuSockets"`
```
CPU should be capitalised on both of these IMO
```suggestion
	// vcpusPerSocket is the number of vCPUs per socket of the VM
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Minimum=1
	VCPUsPerSocket int32 `json:"vcpusPerSocket"`

	// vcpuSockets is the number of vCPU sockets of the VM
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Minimum=1
	VCPUSockets int32 `json:"vcpuSockets"`
```

yanhua121 commented:
Added and refactored the endpoints data for accessing the Prism Central and its Elements in the Infrastructure NutanixPlatformSpec.

yanhua121 commented:
/retest

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -705,7 +705,30 @@ type AlibabaCloudResourceTag struct {
+
+	// prismElements holds an array of endpoint and port data to access the Nutanix
+	// Prism Elements (clusters) of the Nutanix Prism Central
+	// +kubebuilder:validation:Required
+	PrismElements []NutanixPrismEndpoint `json:"prismElements"`
```
How many of these are users expected to supply?

What does each different one represent?

Do they need to be named or have something unique within them?

How will consumers of this field know which of the elements to use?

vnephologist replied:
@JoelSpeed Prism Element (PE) is a single HCI cluster of physical nodes that's managed by a single entity (effectively, API endpoint). These Prism Element instances can then be aggregated and managed by a single Prism Central (PC) instance. We primarily target the PC instance for deployment, but there are some components that cannot (yet) be managed by PC (e.g. CSI storage), so we must also take PE input. For phase one, we are supporting only a single PE cluster for deployment, but in the future, it is possible that a single OCP cluster could span multiple PEs. Also, note that, by default, we ask the user for PE cluster name only, but in theory, it could also have a separate user/pass as well:

  ? Prism Central pc01.ocp.nutanix.com
  ? Prism Central Username admin@ocp.nutanix.com
  ? Prism Central Password [? for help] **********
  ? Cluster (Prism Element) nx01.ocp.nutanix.com

yanhua121 replied:
Added the required name field (to identify it) in the Prism Element Endpoint structure. For this release, we only support one Prism Element (cluster) for an openshift cluster, where all the Nutanix resources (VMs, subnet, volume, etc.) used in the Openshift cluster locate. In the future, we may support the Nutanix resources (VMs, etc.) used in the Openshift cluster can come from multiple Prism Elements (clusters) of the Prism Cental.

JoelSpeed reviewed (commented):
I need to understand a bit more about the new elements thing that's just cropped up, but otherwise I'm pretty happy with this now

yanhua121 commented:
/retest

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -705,7 +705,45 @@ type AlibabaCloudResourceTag struct {
+
+// NutanixPrismElementEndpoint holds the name and endpoint data for a Prism Element (cluster)
+type NutanixPrismElementEndpoint struct {
+	// name is the name of the Prism Element (cluster)
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=256
+	Name string `json:"name"`
```
I'm guessing that this will tie to some object field, so in the machine providerSpec will it have an element name to be deployed to? Likewise, will the CSI storage have a way to configure which element name to be deployed to?

JoelSpeed replied:
It would be good if you could expand this comment to explain that other resources will have a cluster field which will reference a name within this list. The context you've just replied here is very helpful for the user to understand how this value relates to their other resources

yanhua121 replied:
All the configuration details (such as this) will be document in the installation document.

JoelSpeed replied:
This is also user facing documentation and a note should be added here as well IMO. Only needs to be something along the lines of "This value will correspond with the `cluster` field configured on other resources (eg Machines, PVCs, etc)"

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -705,7 +705,45 @@ type AlibabaCloudResourceTag struct {
+	// prismElements holds one or more endpoint address and port data to access the Nutanix
+	// Prism Elements (clusters) of the Nutanix Prism Central. Currently we only support one
+	// Prism Element (cluster) for an openshift cluster, where all the Nutanix resources (VMs, subnet, etc.)
+	// used in the Openshift cluster locate. In the future, we may support the Nutanix resources (VMs, etc.)
+	// used in the Openshift cluster can come from multiple Prism Elements (clusters) of the Prism Cental.
```
Nit: You have mixed the casing of `OpenShift` within this paragraph, you can choose either `openshift` or the preferred `OpenShift`, but you need to make it consistent, this doc string is user facing.

Also, I think this will read better, wdyt?
```suggestion
	// used in the OpenShift cluster are located. In the future, we may support Nutanix resources (VMs, etc.)
	// spread over multiple Prism Elements (clusters) of the Prism Central.
```

JoelSpeed reviewed (commented):
Some small nits on the prism elements stuff, otherwise I think this is good now

/assign @deads2k 
for a second review

deads2k commented on config/v1/types_infrastructure.go:
```diff
@@ -705,7 +705,45 @@ type AlibabaCloudResourceTag struct {
+type NutanixPlatformSpec struct {
+	// prismCentral holds the endpoint address and port to access the Nutanix Prism Central
+	// +kubebuilder:validation:Required
+	PrismCentral NutanixPrismEndpoint `json:"prismCentral"`
```
When a proxy is configured, is this endpoint accessed via the proxy or not accessed via the proxy?

yanhua121 replied:
Nutanix Prism provides proxy support and it's documented here https://portal.nutanix.com/page/documents/details?targetId=Prism-Central-Guide-Prism-v6_0:mul-http-proxy-configure-pc-t.html.

yanhua121 replied:
Added the comment.

deads2k commented on config/v1/types_infrastructure.go:
```diff
@@ -705,7 +705,47 @@ type AlibabaCloudResourceTag struct {
+
+	// endpoint holds the endpoint address and port data of the Prism Element (cluster)
+	// +kubebuilder:validation:Required
+	Endpoint NutanixPrismEndpoint `json:"endpoint"`
```
When a proxy is configured, is this endpoint accessed via the proxy or not accessed via the proxy? My expectation would be NO_PROXY, but that will require some plumbing.

yanhua121 replied:
Nutanix Prism provides proxy support and it's documented here https://portal.nutanix.com/page/documents/details?targetId=Prism-Central-Guide-Prism-v6_0:mul-http-proxy-configure-pc-t.html.

yanhua121 replied:
Added the comment.

yanhua121 commented:
@JoelSpeed @deads2k I addressed all your review comments. Please let us know what to do to speed up approval of this PR? As this PR is the basis for the corresponding changes in the mapi-nutanix and installer repos.

elmiko commented:
i talked with @JoelSpeed and @yanhua121 , and i think we've addressed the final issue about the comment on the `Name` field.

yanhua121 commented:
/retest

elmiko commented:
it looks like this needs an update to the generated files to pick up the comment changes.

JoelSpeed commented:
/lgtm

This project is not a part of the no-ff process, so I will label it appropriately for the docs/qe/px labels

/label px-approved
/label qe-approved
/label docs-approved

JoelSpeed commented:
/lgtm cancel

@yanhua121 I've spoken with @deads2k and clarified with @vnephologist what we are expecting here. If you can cherry-pick https://github.com/JoelSpeed/api/commit/5d7351f5061ef54e8df87a8c9517209fb1bdb616 and then squash all of the commits on this branch down, we are ready to merge

JoelSpeed reviewed (commented):
/lgtm

deads2k commented:
Thanks for the updates.  Looks good

/approve
